log/record.go

```diff
@@ -0,0 +1,183 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+// Copyright 2022 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package log // import "go.opentelemetry.io/otel/log"
+
+import (
+	"errors"
+	"time"
+
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+)
+
+var errUnsafeAddAttrs = errors.New("unsafely called AddAttrs on copy of Record made without using Record.Clone")
+
+// Record TODO: comment.
+// TODO: Add unit tests.
+type Record struct {
+	// TODO: comment.
+	Timestamp time.Time
+
+	// TODO: comment.
+	ObservedTimestamp time.Time
+
+	// TODO: comment.
+	Severity Severity
+
+	// TODO: comment.
+	SeverityText string
+
+	// TODO: comment.
+	Body string
+
+	// The fields below are for optimizing the implementation of
+	// Attributes and AddAttributes.
+
+	// Allocation optimization: an inline array sized to hold
+	// the majority of log calls (based on examination of open-source
+	// code). It holds the start of the list of attributes.
+	front [AttributesInlineCount]attribute.KeyValue
+
+	// The number of attributes in front.
+	nFront int
+
+	// The list of attributes except for those in front.
+	// Invariants:
+	//   - len(back) > 0 iff nFront == len(front)
+	//   - Unused array elements are zero. Used to detect mistakes.
+	back []attribute.KeyValue
+}
+
+// AttributesInlineCount TODO: comment that this should be used
+// by bridges to implement in order to reduce memory allocations.
+const AttributesInlineCount = 5
```

Comment:
This seems like a leak of an implementation detail.

Reply:
This is true. However, I feel that it could be exported as the log bridges could use it to have an implementation that minimizes the heap allocations. You can take a look at the benchmarks and how log bridges are implemented.

Reply:
I think it should be unexported

Reply:
How about adding your concern to "open issues" design doc section?

I think that we should create docs that would advise how to write a high performant log bridge. Something similar (simpler for sure) to https://github.com/golang/example/blob/master/slog-handler-guide/README.md

Reply:
**Me wanting making it exported:**

If we decide to use different internal implementation in future (e.g. because we will learn or there would be new ways to implement it more efficently). Then we can deprecate `AttributesInlineCount`. I think it would be better than expecting users to hard-code `5` and having no hints (other than documentation) if the internals change. I belive that such interals matter 

**Me agreeing with @MrAlias and affraid of uncessary big API surface**:

> We found that over 95% of calls to logging methods pass five or fewer attributes

The implementation should target the hot path and not care so much about the scenarios when more attributes are added. Also these are only attributes added for a single recording. The resource attributes, trace ID, span ID are added separately so they do not count to this `5`.

I am leaning towards unexpecting. I will execute the benchmark and compare the differences in the bridge implementations. If it will be acceptable (which I think it will) then I will describe  "exported AttributesInlineCount" as a rejected proposal. We can always introduce it later.

Reply:
There are only a few more allocs when attributes count > 5.
Without the `if` conditional the for attributes count <= 5 it is even faster.
I think this is the way to go.

@MrAlias Do you agree?

```
$ benchstat old.txt new.txt
goos: linux
goarch: amd64
pkg: go.opentelemetry.io/otel/log/benchmark
cpu: Intel(R) Core(TM) i9-10885H CPU @ 2.40GHz
                        │    old.txt    │                new.txt                 │
                        │    sec/op     │     sec/op      vs base                │
Emit/noop/no_attrs-16      22.92n ±  5%    20.77n ±   3%   -9.40% (p=0.000 n=10)
Emit/noop/3_attrs-16       79.52n ±  6%    71.67n ±   2%   -9.87% (p=0.000 n=10)
Emit/noop/5_attrs-16      114.05n ±  4%    98.41n ±   2%  -13.72% (p=0.000 n=10)
Emit/noop/10_attrs-16      338.2n ±  7%    286.8n ±   2%  -15.21% (p=0.000 n=10)
Emit/noop/40_attrs-16      1.359µ ±  7%    1.064µ ±   2%  -21.68% (p=0.000 n=10)
Emit/writer/no_attrs-16    166.2n ±  4%    133.7n ±   1%  -19.61% (p=0.000 n=10)
Emit/writer/3_attrs-16     599.8n ±  5%    488.3n ±   1%  -18.58% (p=0.000 n=10)
Emit/writer/5_attrs-16     729.6n ±  3%    599.0n ±   2%  -17.89% (p=0.000 n=10)
Emit/writer/10_attrs-16    1.544µ ± 26%    1.202µ ±   2%  -22.15% (p=0.000 n=10)
Emit/writer/40_attrs-16    5.252µ ±  5%    4.352µ ±   3%  -17.14% (p=0.000 n=10)
Slog/no_attrs-16           451.1n ±  5%    388.6n ±   3%  -13.87% (p=0.000 n=10)
Slog/3_attrs-16            671.5n ±  3%    602.0n ±   2%  -10.34% (p=0.000 n=10)
Slog/5_attrs-16            796.2n ±  4%    757.8n ±   2%   -4.83% (p=0.001 n=10)
Slog/10_attrs-16           1.598µ ±  6%    1.681µ ±   2%   +5.23% (p=0.011 n=10)
Slog/40_attrs-16           4.531µ ±  5%    5.619µ ±   3%  +24.01% (p=0.000 n=10)
Logr/no_attrs-16           36.25n ± 10%    30.69n ±   4%  -15.33% (p=0.000 n=10)
Logr/3_attrs-16            225.7n ±  8%    241.8n ± 153%        ~ (p=0.061 n=10)
Logr/5_attrs-16            322.9n ±  3%    558.9n ±  31%  +73.11% (p=0.000 n=10)
Logr/10_attrs-16           864.1n ±  3%   1038.5n ±  31%  +20.18% (p=0.000 n=10)
Logr/40_attrs-16           3.400µ ±  5%    4.419µ ±   5%  +29.96% (p=0.000 n=10)
geomean                    494.5n          471.2n          -4.70%

                        │    old.txt     │                new.txt                 │
                        │      B/op      │     B/op      vs base                  │
Emit/noop/no_attrs-16       0.000 ± 0%       0.000 ± 0%        ~ (p=1.000 n=10) ¹
Emit/noop/3_attrs-16        0.000 ± 0%       0.000 ± 0%        ~ (p=1.000 n=10) ¹
Emit/noop/5_attrs-16        0.000 ± 0%       0.000 ± 0%        ~ (p=1.000 n=10) ¹
Emit/noop/10_attrs-16       320.0 ± 0%       320.0 ± 0%        ~ (p=1.000 n=10) ¹
Emit/noop/40_attrs-16     2.250Ki ± 0%     2.250Ki ± 0%        ~ (p=1.000 n=10) ¹
Emit/writer/no_attrs-16     16.00 ± 0%       16.00 ± 0%        ~ (p=1.000 n=10) ¹
Emit/writer/3_attrs-16      48.00 ± 0%       48.00 ± 0%        ~ (p=1.000 n=10) ¹
Emit/writer/5_attrs-16      48.00 ± 0%       48.00 ± 0%        ~ (p=1.000 n=10) ¹
Emit/writer/10_attrs-16     408.0 ± 0%       408.0 ± 0%        ~ (p=1.000 n=10) ¹
Emit/writer/40_attrs-16   2.570Ki ± 0%     2.570Ki ± 0%        ~ (p=1.000 n=10) ¹
Slog/no_attrs-16            0.000 ± 0%       0.000 ± 0%        ~ (p=1.000 n=10) ¹
Slog/3_attrs-16             0.000 ± 0%       0.000 ± 0%        ~ (p=1.000 n=10) ¹
Slog/5_attrs-16             0.000 ± 0%       0.000 ± 0%        ~ (p=1.000 n=10) ¹
Slog/10_attrs-16          1.141Ki ± 0%     1.141Ki ± 0%        ~ (p=1.000 n=10) ¹
Slog/40_attrs-16          6.250Ki ± 0%     9.312Ki ± 0%  +49.00% (p=0.000 n=10)
Logr/no_attrs-16            0.000 ± 0%       0.000 ± 0%        ~ (p=1.000 n=10) ¹
Logr/3_attrs-16             128.0 ± 0%       128.0 ± 0%        ~ (p=1.000 n=10) ¹
Logr/5_attrs-16             208.0 ± 0%       208.0 ± 0%        ~ (p=1.000 n=10) ¹
Logr/10_attrs-16          1.344Ki ± 0%     1.344Ki ± 0%        ~ (p=1.000 n=10) ¹
Logr/40_attrs-16          6.500Ki ± 0%     9.562Ki ± 0%  +47.12% (p=0.000 n=10)
geomean                                ²                  +4.00%                ²
¹ all samples are equal
² summaries must be >0 to compute geomean

                        │   old.txt    │                new.txt                │
                        │  allocs/op   │ allocs/op   vs base                   │
Emit/noop/no_attrs-16     0.000 ± 0%     0.000 ± 0%         ~ (p=1.000 n=10) ¹
Emit/noop/3_attrs-16      0.000 ± 0%     0.000 ± 0%         ~ (p=1.000 n=10) ¹
Emit/noop/5_attrs-16      0.000 ± 0%     0.000 ± 0%         ~ (p=1.000 n=10) ¹
Emit/noop/10_attrs-16     1.000 ± 0%     1.000 ± 0%         ~ (p=1.000 n=10) ¹
Emit/noop/40_attrs-16     1.000 ± 0%     1.000 ± 0%         ~ (p=1.000 n=10) ¹
Emit/writer/no_attrs-16   1.000 ± 0%     1.000 ± 0%         ~ (p=1.000 n=10) ¹
Emit/writer/3_attrs-16    4.000 ± 0%     4.000 ± 0%         ~ (p=1.000 n=10) ¹
Emit/writer/5_attrs-16    4.000 ± 0%     4.000 ± 0%         ~ (p=1.000 n=10) ¹
Emit/writer/10_attrs-16   8.000 ± 0%     8.000 ± 0%         ~ (p=1.000 n=10) ¹
Emit/writer/40_attrs-16   26.00 ± 0%     26.00 ± 0%         ~ (p=1.000 n=10) ¹
Slog/no_attrs-16          0.000 ± 0%     0.000 ± 0%         ~ (p=1.000 n=10) ¹
Slog/3_attrs-16           0.000 ± 0%     0.000 ± 0%         ~ (p=1.000 n=10) ¹
Slog/5_attrs-16           0.000 ± 0%     0.000 ± 0%         ~ (p=1.000 n=10) ¹
Slog/10_attrs-16          3.000 ± 0%     5.000 ± 0%   +66.67% (p=0.000 n=10)
Slog/40_attrs-16          3.000 ± 0%     8.000 ± 0%  +166.67% (p=0.000 n=10)
Logr/no_attrs-16          0.000 ± 0%     0.000 ± 0%         ~ (p=1.000 n=10) ¹
Logr/3_attrs-16           4.000 ± 0%     4.000 ± 0%         ~ (p=1.000 n=10) ¹
Logr/5_attrs-16           5.000 ± 0%     5.000 ± 0%         ~ (p=1.000 n=10) ¹
Logr/10_attrs-16          11.00 ± 0%     13.00 ± 0%   +18.18% (p=0.000 n=10)
Logr/40_attrs-16          35.00 ± 0%     40.00 ± 0%   +14.29% (p=0.000 n=10)
geomean                              ²                 +9.37%                ²
¹ all samples are equal
² summaries must be >0 to compute geomean
```

```diff
$ git --no-pager diff
diff --git a/log/benchmark/logr_test.go b/log/benchmark/logr_test.go
index 4af9cb75..b4f637c2 100644
--- a/log/benchmark/logr_test.go
+++ b/log/benchmark/logr_test.go
@@ -55,27 +55,13 @@ func (s *logrSink) Info(level int, msg string, keysAndValues ...any) {
                panic("key without a value")
        }
        kvCount := len(keysAndValues) / 2
-       if kvCount > log.AttributesInlineCount {
-               attrs := make([]attribute.KeyValue, 0, kvCount)
-               for i := 0; i < kvCount; i++ {
-                       k, ok := keysAndValues[i*2].(string)
-                       if !ok {
-                               panic("key is not a string")
-                       }
-                       kv := convertKV(k, keysAndValues[i*2+1])
-                       attrs = append(attrs, kv)
-               }
-               record.AddAttributes(attrs...)
-       } else {
-               // special case that avoids heap allocations (hot path)
-               for i := 0; i < kvCount; i++ {
-                       k, ok := keysAndValues[i*2].(string)
-                       if !ok {
-                               panic("key is not a string")
-                       }
-                       kv := convertKV(k, keysAndValues[i*2+1])
-                       record.AddAttributes(kv)
+       for i := 0; i < kvCount; i++ {
+               k, ok := keysAndValues[i*2].(string)
+               if !ok {
+                       panic("key is not a string")
                }
+               kv := convertKV(k, keysAndValues[i*2+1])
+               record.AddAttributes(kv)
        }
 
        s.Logger.Emit(ctx, record)
diff --git a/log/benchmark/slog_test.go b/log/benchmark/slog_test.go
index df1b4b88..d150b008 100644
--- a/log/benchmark/slog_test.go
+++ b/log/benchmark/slog_test.go
@@ -42,20 +42,10 @@ func (h *slogHandler) Handle(_ context.Context, r slog.Record) error {
 
        record := log.Record{Timestamp: r.Time, Severity: lvl, Body: r.Message}
 
-       if r.NumAttrs() > log.AttributesInlineCount {
-               attrs := make([]attribute.KeyValue, 0, r.NumAttrs())
-               r.Attrs(func(a slog.Attr) bool {
-                       attrs = append(attrs, convertAttr(a))
-                       return true
-               })
-               record.AddAttributes(attrs...)
-       } else {
-               // special case that avoids heap allocations (hot path)
-               r.Attrs(func(a slog.Attr) bool {
-                       record.AddAttributes(convertAttr(a))
-                       return true
-               })
-       }
+       r.Attrs(func(a slog.Attr) bool {
+               record.AddAttributes(convertAttr(a))
+               return true
+       })
 
        h.Logger.Emit(context.Background(), record)
        return nil
diff --git a/log/record.go b/log/record.go
index 7f82e309..4365c16b 100644
--- a/log/record.go
+++ b/log/record.go
@@ -41,7 +41,7 @@ type Record struct {
        // Allocation optimization: an inline array sized to hold
        // the majority of log calls (based on examination of open-source
        // code). It holds the start of the list of attributes.
-       front [AttributesInlineCount]attribute.KeyValue
+       front [attributesInlineCount]attribute.KeyValue
 
        // The number of attributes in front.
        nFront int
@@ -53,9 +53,7 @@ type Record struct {
        back []attribute.KeyValue
 }
 
-// AttributesInlineCount TODO: comment that this should be used
-// by bridges to implement in order to reduce memory allocations.
-const AttributesInlineCount = 5
+const attributesInlineCount = 5
 
 // Severity TODO: comment.
 type Severity int
```